File: 2015/c++/6/main-part-1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>

void get_action(std::string line, std::string &action, int &startx, int &starty, int &endx, int &endy)
{
    std::string temp;
    char comma;
    std::stringstream ss(line);
    if (line[0] == 't' && line[1] == 'u')
        ss >> temp >> action >> startx >> comma >> starty >> temp >> endx >> comma >> endy;
    else
        ss >> action >> startx >> comma >> starty >> temp >> endx >> comma >> endy;
}

int main(int argc, char **argv)
{
    if (argc < 2)
    {
        std::cout << "Usage: " << argv[0] << " <input.txt>" << std::endl;
        return EXIT_FAILURE;
    }

    std::string filename = argv[1];
    std::fstream file(filename, std::ios::in);

    if (!file.is_open())
    {
        std::cout << "Error: Unable to open file " << filename << std::endl;
        return EXIT_FAILURE;
    }

    std::string line;
    bool lights[1000 * 1000] = {false};

    while (std::getline(file, line))
    {
        std::string action;
        int startx, starty, endx, endy;

        get_action(line, action, startx, starty, endx, endy);

        for (int i = startx; i <= endx; i++)
        {
            for (int j = starty; j <= endy; j++)
            {
                if (action == "on")
                    lights[i * 1000 + j] = true;
                else if (action == "off")
                    lights[i * 1000 + j] = false;
                else
                    lights[i * 1000 + j] = !lights[i * 1000 + j];
            }
        }
    }

    file.close();

    int count = 0;
    for (int i = 0; i < 1000 * 1000; i++)
        count += lights[i];

    std::cout << "Part 1: " << count << std::endl;

    return EXIT_SUCCESS;
}

File: 2015/go/6/main-part-1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func get_action(line string, action *string, startx *int, starty *int, endx *int, endy *int) {
	if line[0:2] == "tu" {
		fmt.Sscanf(line, "turn %s %d,%d through %d,%d", action, startx, starty, endx, endy)
	} else {
		fmt.Sscanf(line, "%s %d,%d through %d,%d", action, startx, starty, endx, endy)
	}
}

func main() {
	args := os.Args[1:]

	if len(args) < 1 {
		fmt.Println("Usage: go run main.go <input.txt>")
		return
	}

	var filename string = args[0]

	file, err := os.Open(filename)

	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	defer file.Close()

	var lights [1000 * 1000]bool

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		var action string
		var startx, starty, endx, endy int
		get_action(line, &action, &startx, &starty, &endx, &endy)

		for x := startx; x <= endx; x++ {
			for y := starty; y <= endy; y++ {
				index := x*1000 + y
				if action == "on" {
					lights[index] = true
				} else if action == "off" {
					lights[index] = false
				} else if action == "toggle" {
					lights[index] = !lights[index]
				}
			}
		}
	}

	// Count the number of lights that are on
	count := 0
	for i := 0; i < 1000*1000; i++ {
		if lights[i] {
			count++
		}
	}

	fmt.Println("Part 1:", count)

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}
}
